### PR TITLE
added gunbag for DMRs

### PR DESCRIPTION
**Pull request by ir0n1e**
Gunbag for DMRs
~~Known issues: Mass calculation of full gunbag - cause of ArmA restrictions~~
Screenshots (tan)
http://imgur.com/a/r9l36

### CODE REVIEW

**Comment by jonpas** [comment]
Replace tabs with 4 spaces. 

Can you also give a broader description of the feature?

**Comment by BaerMitUmlaut** [comment]
Looks like this adds an interaction to add a weapon virtually to a gunbag backpack.

This won't work with weapons that have scopes by default which were then removed by a script.

Additionally I don't see the need for this at all. You can already put weapons in your backpack. Adding a larger backpack to the game would have the same effect without the potential for bugs and issues.

**Comment by ir0n1e** [comment]
@BaerMitUmlaut 
Part1: correct
Part2: Possible but could not repuduse in tests
Part3: Your opinion

**Comment by bux578** [comment]
@ir0n1e 
Part 2 isn't possible in Arma 1.58 with the commands that are available.

**Comment by BaerMitUmlaut** [comment]
Part 3 is not an opinion. What advantages does this have that a large backpack doesn't have? I see the following disadvantages:

- breaks with some weapons because of how attachments work
- removes the effect of weight of weapons in the backpack and is independent of what weapon is in the backpack (an SMG would weigh as much as an MMG)
- removes the interaction of opening the backpack and switching out your weapon manually (realism)

**Comment by ir0n1e** [comment]
@bux578 Jabb, but I read current weapon config with EFUNC(common,getWeaponState)
and rebuilding the weapon to this state.

@BaerMitUmlaut you are right, but you can't store a G82 in any backpack (realism).
My prefered solution is a weapon in the inventory of the gunbag, but it's not possible to buildup the weapon with all attachments in backpack.

**Comment by BaerMitUmlaut** [comment]
This would work with making a new backpack that is simply larger.

**Comment by ir0n1e** [comment]
The G82 has more then 60 inches, the largest backpack has less then 30 inches. You have to dismatle the weapon to store it in a backpack. (realism
)

**Comment by BaerMitUmlaut** [comment]
>This would work with making a new backpack that is simply larger.

**Comment by ir0n1e** [comment]
Knowing that. Your comment was realism.

**Comment by ir0n1e** [comment]
@commy2 needing your opinion, caues it's your initial idea!

**Comment by commy2** [comment]
>This would work with making a new backpack that is simply larger.

Not true. "Long rifles" (anything that isn't a 5.56mm or 6.5mm by vanilla standards) cannot be put inside a bagpack, regardless of how little `mass` they have or how much cargo capacity the bagpack has.

```
    class Rifle : RifleCore {
        class WeaponSlotsInfo {
            allowedSlots[] = {901};
        };
    };
```

```
    class Rifle_Long_Base_F : Rifle_Base_F {
        class WeaponSlotsInfo : WeaponSlotsInfo {
            allowedSlots[] = {};
        };
    };
```

- 701 vest
- 801 uniform
- 901 backpack

**Comment by BaerMitUmlaut** [comment]
Why not change the configs accordingly then?

**Comment by commy2** [comment]
>breaks with some weapons because of how attachments work

I don't see how. 

Store weapon info:
While it's true that pre-equipped weapons are reported as their class names with attachments, the `primaryWeaponItems` command still reports only the attachments currently there, including those added automatically.

Restore weapon:
You use `addWeapon`, ´removeAllPrimaryWeaponItems` and then `addPrimaryWeaponItem` and can fully restore a primary weapon. The problem comes with launchers, because the commands are bugged for those.

Pre-quipped weapons are not a problem.

**Comment by BaerMitUmlaut** [comment]
That's not the case right now though.

**Comment by commy2** [comment]
>Why not change the configs accordingly then?

Then you could put (multiple of) those weapons inside even the smallest backpacks, even though they shouldn't have the space for it. It's just the cheap vanilla inventory system.

**Comment by commy2** [comment]
>That's not the case right now though.

I haven't checked the PR, but it is possible to achive (with primary weapons)

**Comment by ir0n1e** [comment]
Test it and you will see it works. Script stores weapon, and attachements. If you pull out the weapon all attacements will be build up.
Problem is mass of gunbag, cause we can't build up the weapon in a vehicle.

**Comment by commy2** [comment]
>Problem is mass of gunbag, cause we can't build up the weapon in a vehicle.

while this is true, you could still adjust the "loadCoef" via `setUnitTrait` to create some kind of "virtual mass". I haven't tested the command for bugs though ...

**Comment by ir0n1e** [comment]
Will test that, but first time read this command. Have tested with "setmass", but is's not working for player vehicle.

**Comment by ir0n1e** [comment]
@commy2 setUnitTrait sets unit specialisation..? Do you have time for a TS conversation today?

**Comment by BaerMitUmlaut** [comment]
Try adding weapon that has attachments by default for example `arifle_TRG21_ARCO_pointer_F`. Now drop all attachments, put it in the backpack and take it out again. I'm pretty sure this creates new attachments right now. I cannot try it right now.

**Comment by ir0n1e** [comment]
@BaerMitUmlaut jabb, the script will working so, but it's nothing that I can change.

**Comment by commy2** [comment]
>Do you have time for a TS conversation today?
Yes, `1700B`

>setUnitTrait sets unit specialisation..?
I think that was what they were going for, but they ended up also enabling editing some other numeric values.

https://community.bistudio.com/wiki/setUnitTrait

**Comment by ir0n1e** [comment]
@commy2 1700 thats a date!

**Comment by BaerMitUmlaut** [comment]
You need to use [removeAllPrimaryWeaponItems](https://community.bistudio.com/wiki/removeAllPrimaryWeaponItems) first.

**Comment by ir0n1e** [comment]
@BaerMitUmlaut Will do that

**Comment by jonpas** [comment]
```
ERROR: Tab detected at addons/gunbag/functions/fnc_canInteract.sqf Line number: 24
ERROR: Tab detected at addons/gunbag/functions/fnc_canInteract.sqf Line number: 28
ERROR: Tab detected at addons/gunbag/functions/fnc_offGunbag.sqf Line number: 29
ERROR: Tab detected at addons/gunbag/functions/fnc_offGunbag.sqf Line number: 39
ERROR: Tab detected at addons/gunbag/functions/fnc_status.sqf Line number: 20
ERROR: Tab detected at addons/gunbag/functions/fnc_status.sqf Line number: 22
ERROR: Tab detected at addons/gunbag/functions/fnc_status.sqf Line number: 23
ERROR: Tab detected at addons/gunbag/functions/fnc_status.sqf Line number: 24
ERROR: Tab detected at addons/gunbag/functions/fnc_switchBackpack.sqf Line number: 28
ERROR: Tab detected at addons/gunbag/functions/fnc_switchBackpack.sqf Line number: 32
```

**Comment by ir0n1e** [comment]
It's a bot opinion. Say me where is the tab in addons/gunbag/functions/fnc_canInteract.sqf Line number: 24

File: addons/gunbag/functions/fnc_canInteract.sqf [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+params ["_unit","_target"];
+
+private ["_result","_gunbag"];
+
+_result = -1;
+_gunbag = backpackContainer _target;
+
+if(count (_gunbag getVariable [QGVAR(GunbagWeapon),[]]) <= 0 && {primaryWeapon _unit != ""} && {getNumber (configFile >> "CfgWeapons" >> primaryWeapon _unit  >> QGVAR(allowGunbag)) == 1}) then {
+	_result = 0;
```
Comment (jonpas) [inline addons/gunbag/functions/fnc_canInteract.sqf]: Tab

**Comment by ir0n1e** [comment]
seeing a tab in line 25 but not in 24.

**Comment by Glowbal** [comment]
Hi @ir0n1e,

Our validating script started counting at 0 instead of 1. I've changed it on our master branch to start counting at 1 now. 

Meanwhile, I am interested in this PR - however could you please update it to match our style guide? That is 4 spaces for indentation and no tabs. This is for all .sqf, .hpp, .cpp files at the least. Doing this helps us keep a clean and consistent code base. A simple search and replace all tabs should do the trick. Let us know if you need assistance any of this.

Thanks.

**Comment by ir0n1e** [comment]
@Glowbal jab will do that.

**Comment by ir0n1e** [comment]
<german jab> == <english yeah>

**Comment by commy2** [comment]
@ir0n1e I finished the virtual load framework. just use:

`[_unit, backpackContainer _unit, _mass] call EFUNC(movement,addLoadToUnitContainer);`

Use a negative mass when removing the virtual rifle from the backpack.

**Comment by ir0n1e** [comment]
kk thanks

**Comment by ir0n1e** [comment]
@commy2 Mass calculation: Which mass i have to pass (config mass or Kg/Pounds)?
Virtual mass is not recognized by EFUNC(movement,getWeight)!

**Comment by esteldunedain** [comment]
> Virtual mass is not recognized by EFUNC(movement,getWeight)!

Good point.

**Comment by BaerMitUmlaut** [comment]
So I tried to confirm this @commy2:
>Then you could put (multiple of) those weapons inside even the smallest backpacks, even though they shouldn't have the space for it. It's just the cheap vanilla inventory system.

Config changes:
```
class Rifle_Long_Base_F: Rifle_Base_F {
    class WeaponSlotsInfo: WeaponSlotsInfo {
        allowedSlots[] = {901};
    };
};
```
and
```
class B_Carryall_Base: Bag_Base {
    maximumLoad = 500;
};
```

Result:
img


Where's the issue? I really like the gunbag model btw. I'd say make it have a large inventory space and add it to ACEX?

**Comment by jokoho48** [comment]
```
class B_Carryall_Base: Bag_Base {
    maximumLoad = 500;
};
```
i think this is a horrible idea dont change Weight or Capacity values

**Comment by BaerMitUmlaut** [comment]
Not saying we should change the load value of the Carryall. I'm saying we should make this a new backpack with a similarily sized `maximumLoad`.

**Comment by jokoho48** [comment]
but @BaerMitUmlaut what is happening if someone want to put RHS or CUP or BW weapons like this in there backpack and its dont work... they came to use and Blame us that we broke it.

**Comment by BaerMitUmlaut** [comment]
If they inherit from `Rifle_Long_Base_F` as they probably should there are no issues. If they don't and manually disable adding them to backpacks, it's clearly their fault. I don't see the issue with that.

**Comment by ir0n1e** [comment]
At this time script checks a config value. Like swapbarrol. The modders have to add this value in there ace_comp config. If you check "Rifle_Long_Base_F", LMGs and MMGs will fit into gunbag.

**Comment by BaerMitUmlaut** [comment]
What's the problem with putting an MMG in the gunbag? We even have deployable HMGs in Arma that come in backpacks. Why not MMGs now?

**Comment by ir0n1e** [comment]
Don't have a major problem with that, but it schould be a bag for sniperrifles in my opinion. In real will you carray a Barrett or a M24 in this bag.

**Comment by ir0n1e** [comment]
> @commy2 Mass calculation: Which mass i have to pass (config mass or Kg/Pounds)?
> Virtual mass is not recognized by EFUNC(movement,getWeight)!

**Comment by ir0n1e** [comment]
KK, thik I have to pass config mass, cause of #3779

File: addons/gunbag/functions/fnc_offGunbag.sqf [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+};
+
+_weapon = (_state select 0) select 0;
+_items = (_state select 1);
+
+_unit addWeapon _weapon;
+removeAllPrimaryWeaponItems _unit;
+
+{
+  _unit addWeaponItem [_weapon, _x];
```
Comment (commy2) [inline addons/gunbag/functions/fnc_offGunbag.sqf]: 4 spaces indent

Comment (jonpas) [inline addons/gunbag/functions/fnc_offGunbag.sqf]: Everywhere actually. :)

**Comment by jokoho48** [comment]
ok .. this time in the right PR... fu github why did you do this...

i just took a look at the model and found 2 thing that you should improve, the Polycount in this parts of the model is way to high
http://puu.sh/oNmPh/b985503f8b.jpg
http://puu.sh/oNmTj/ad8adb18ac.png

**Comment by ir0n1e** [comment]
@jokoho48 will see what I can do. But I hate bumpmapping.

**Comment by jokoho48** [comment]
i think that can be dont without any normal

**Comment by ir0n1e** [comment]
I'm done. I Think, everything should pass ACE3 checks.

**Comment by ir0n1e** [comment]
Merge status?

**Comment by SzwedzikPL** [comment]
Gunbag for weapon carried on launcher slot would be an awesome feature

**Pull request merged**